File: coreengine/kernel/services.go
// Package kernel provides service registry - IPC coordinator equivalent.
//
// Features:
//   - Service registration and discovery
//   - Health tracking
//   - Load balancing
//   - Dispatch routing
package kernel

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// =============================================================================
// Service Status
// =============================================================================

// ServiceStatus represents the health status of a service.
type ServiceStatus string

const (
	// ServiceStatusHealthy indicates the service is healthy and accepting requests.
	ServiceStatusHealthy ServiceStatus = "healthy"
	// ServiceStatusDegraded indicates the service is running but with reduced capacity.
	ServiceStatusDegraded ServiceStatus = "degraded"
	// ServiceStatusUnhealthy indicates the service is not accepting requests.
	ServiceStatusUnhealthy ServiceStatus = "unhealthy"
	// ServiceStatusUnknown indicates the service status is unknown.
	ServiceStatusUnknown ServiceStatus = "unknown"
)

// =============================================================================
// Service Types
// =============================================================================

// ServiceType constants for classifying services.
const (
	// ServiceTypeFlow is for flow orchestration services.
	ServiceTypeFlow = "flow"
	// ServiceTypeWorker is for worker/execution services.
	ServiceTypeWorker = "worker"
	// ServiceTypeVertical is for vertical/domain-specific services.
	ServiceTypeVertical = "vertical"
	// ServiceTypeInference is for ML inference services (embeddings, NLI, etc).
	ServiceTypeInference = "inference"
)

// =============================================================================
// Service Descriptor
// =============================================================================

// ServiceInfo describes a registered service.
type ServiceInfo struct {
	// Identity
	Name        string `json:"name"`
	ServiceType string `json:"service_type"` // "flow", "worker", "vertical"
	Version     string `json:"version"`

	// Capabilities
	Capabilities []string `json:"capabilities"`

	// Capacity
	MaxConcurrent int `json:"max_concurrent"`
	CurrentLoad   int `json:"current_load"`

	// Health
	Status          ServiceStatus `json:"status"`
	LastHealthCheck time.Time     `json:"last_health_check"`

	// Metadata
	Metadata map[string]any `json:"metadata,omitempty"`
}

// NewServiceInfo creates a new service descriptor.
func NewServiceInfo(name, serviceType string) *ServiceInfo {
	return &ServiceInfo{
		Name:            name,
		ServiceType:     serviceType,
		Version:         "1.0.0",
		Capabilities:    []string{},
		MaxConcurrent:   10,
		CurrentLoad:     0,
		Status:          ServiceStatusHealthy,
		LastHealthCheck: time.Now().UTC(),
	}
}

// CanAccept checks if the service can accept more load.
func (s *ServiceInfo) CanAccept() bool {
	return s.Status == ServiceStatusHealthy && s.CurrentLoad < s.MaxConcurrent
}

// IsHealthy checks if the service is healthy.
func (s *ServiceInfo) IsHealthy() bool {
	return s.Status == ServiceStatusHealthy || s.Status == ServiceStatusDegraded
}

// Clone creates a copy of the service info.
func (s *ServiceInfo) Clone() *ServiceInfo {
	clone := &ServiceInfo{
		Name:            s.Name,
		ServiceType:     s.ServiceType,
		Version:         s.Version,
		MaxConcurrent:   s.MaxConcurrent,
		CurrentLoad:     s.CurrentLoad,
		Status:          s.Status,
		LastHealthCheck: s.LastHealthCheck,
	}
	if s.Capabilities != nil {
		clone.Capabilities = make([]string, len(s.Capabilities))
		copy(clone.Capabilities, s.Capabilities)
	}
	if s.Metadata != nil {
		clone.Metadata = make(map[string]any, len(s.Metadata))
		for k, v := range s.Metadata {
			clone.Metadata[k] = v
		}
	}
	return clone
}

// =============================================================================
// Dispatch Target
// =============================================================================

// DispatchTarget represents a target for request dispatch.
type DispatchTarget struct {
	ServiceName    string             `json:"service_name"`
	Method         string             `json:"method"`
	Priority       SchedulingPriority `json:"priority"`
	TimeoutSeconds int                `json:"timeout_seconds"`
	RetryCount     int                `json:"retry_count"`
	MaxRetries     int                `json:"max_retries"`
}

// CanRetry checks if the dispatch can be retried.
func (d *DispatchTarget) CanRetry() bool {
	return d.RetryCount < d.MaxRetries
}

// IncrementRetry increments the retry count.
func (d *DispatchTarget) IncrementRetry() {
	d.RetryCount++
}

// =============================================================================
// Dispatch Result
// =============================================================================

// DispatchResult represents the result of a dispatch operation.
type DispatchResult struct {
	Success  bool           `json:"success"`
	Error    string         `json:"error,omitempty"`
	Data     map[string]any `json:"data,omitempty"`
	Duration time.Duration  `json:"duration"`
	Retries  int            `json:"retries"`
}

// =============================================================================
// Service Handler
// =============================================================================

// ServiceHandler is a function that handles service requests.
type ServiceHandler func(ctx context.Context, target *DispatchTarget, data map[string]any) (*DispatchResult, error)

// =============================================================================
// Service Registry
// =============================================================================

// ServiceRegistry manages service registration, discovery, and dispatch.
// Thread-safe implementation for tracking registered services.
//
// Usage:
//
//	registry := NewServiceRegistry(nil)
//
//	// Register a service
//	registry.RegisterService(NewServiceInfo("flow_service", "flow"))
//
//	// Register a handler
//	registry.RegisterHandler("flow_service", myHandler)
//
//	// Dispatch a request
//	result := registry.Dispatch(ctx, target, data)
type ServiceRegistry struct {
	logger Logger

	// Service registry
	services map[string]*ServiceInfo

	// Dispatch handlers
	handlers map[string]ServiceHandler

	mu sync.RWMutex
}

// NewServiceRegistry creates a new service registry.
func NewServiceRegistry(logger Logger) *ServiceRegistry {
	return &ServiceRegistry{
		logger:   logger,
		services: make(map[string]*ServiceInfo),
		handlers: make(map[string]ServiceHandler),
	}
}

// =============================================================================
// Service Registration
// =============================================================================

// RegisterService registers a service with the registry.
// Returns true if registration succeeded, false if service already exists.
func (sr *ServiceRegistry) RegisterService(info *ServiceInfo) bool {
	sr.mu.Lock()
	defer sr.mu.Unlock()

	if _, exists := sr.services[info.Name]; exists {
		if sr.logger != nil {
			sr.logger.Warn("service_already_registered", "service_name", info.Name)
		}
		return false
	}

	sr.services[info.Name] = info

	if sr.logger != nil {
		sr.logger.Info("service_registered",
			"service_name", info.Name,
			"service_type", info.ServiceType,
			"capabilities", info.Capabilities,
		)
	}

	return true
}

// UnregisterService unregisters a service.
// Returns true if unregistration succeeded.
func (sr *ServiceRegistry) UnregisterService(serviceName string) bool {
	sr.mu.Lock()
	defer sr.mu.Unlock()

	if _, exists := sr.services[serviceName]; !exists {
		return false
	}

	delete(sr.services, serviceName)
	delete(sr.handlers, serviceName)

	if sr.logger != nil {
		sr.logger.Info("service_unregistered", "service_name", serviceName)
	}

	return true
}

// RegisterHandler registers a handler for a service.
func (sr *ServiceRegistry) RegisterHandler(serviceName string, handler ServiceHandler) {
	sr.mu.Lock()
	defer sr.mu.Unlock()

	sr.handlers[serviceName] = handler

	if sr.logger != nil {
		sr.logger.Debug("handler_registered", "service_name", serviceName)
	}
}

// =============================================================================
// Service Discovery
// =============================================================================

// GetService gets a service by name.
func (sr *ServiceRegistry) GetService(serviceName string) *ServiceInfo {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	if svc, exists := sr.services[serviceName]; exists {
		return svc.Clone()
	}
	return nil
}

// ListServices lists all registered services matching criteria.
func (sr *ServiceRegistry) ListServices(serviceType string, healthyOnly bool) []*ServiceInfo {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	var result []*ServiceInfo
	for _, svc := range sr.services {
		if serviceType != "" && svc.ServiceType != serviceType {
			continue
		}
		if healthyOnly && !svc.IsHealthy() {
			continue
		}
		result = append(result, svc.Clone())
	}
	return result
}

// GetServiceNames returns all registered service names.
func (sr *ServiceRegistry) GetServiceNames() []string {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	names := make([]string, 0, len(sr.services))
	for name := range sr.services {
		names = append(names, name)
	}
	return names
}

// HasService checks if a service is registered.
func (sr *ServiceRegistry) HasService(serviceName string) bool {
	sr.mu.RLock()
	defer sr.mu.RUnlock()
	_, exists := sr.services[serviceName]
	return exists
}

// HasHandler checks if a handler is registered for a service.
func (sr *ServiceRegistry) HasHandler(serviceName string) bool {
	sr.mu.RLock()
	defer sr.mu.RUnlock()
	_, exists := sr.handlers[serviceName]
	return exists
}

// =============================================================================
// Health Management
// =============================================================================

// UpdateHealth updates the health status of a service.
func (sr *ServiceRegistry) UpdateHealth(serviceName string, status ServiceStatus) bool {
	sr.mu.Lock()
	defer sr.mu.Unlock()

	svc, exists := sr.services[serviceName]
	if !exists {
		return false
	}

	svc.Status = status
	svc.LastHealthCheck = time.Now().UTC()

	if sr.logger != nil {
		sr.logger.Debug("service_health_updated",
			"service_name", serviceName,
			"status", string(status),
		)
	}

	return true
}

// GetHealthyCount returns the count of healthy services.
func (sr *ServiceRegistry) GetHealthyCount() int {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	count := 0
	for _, svc := range sr.services {
		if svc.IsHealthy() {
			count++
		}
	}
	return count
}

// =============================================================================
// Load Management
// =============================================================================

// IncrementLoad increments the current load for a service.
// Returns false if service not found or at capacity.
func (sr *ServiceRegistry) IncrementLoad(serviceName string) bool {
	sr.mu.Lock()
	defer sr.mu.Unlock()

	svc, exists := sr.services[serviceName]
	if !exists || !svc.CanAccept() {
		return false
	}

	svc.CurrentLoad++
	return true
}

// DecrementLoad decrements the current load for a service.
func (sr *ServiceRegistry) DecrementLoad(serviceName string) {
	sr.mu.Lock()
	defer sr.mu.Unlock()

	svc, exists := sr.services[serviceName]
	if !exists {
		return
	}

	if svc.CurrentLoad > 0 {
		svc.CurrentLoad--
	}
}

// GetLoad returns the current load for a service.
func (sr *ServiceRegistry) GetLoad(serviceName string) int {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	svc, exists := sr.services[serviceName]
	if !exists {
		return -1
	}
	return svc.CurrentLoad
}

// =============================================================================
// Dispatch
// =============================================================================

// Dispatch dispatches a request to a service.
func (sr *ServiceRegistry) Dispatch(
	ctx context.Context,
	target *DispatchTarget,
	data map[string]any,
) *DispatchResult {
	startTime := time.Now()

	// Check if service exists
	sr.mu.RLock()
	svc, exists := sr.services[target.ServiceName]
	handler := sr.handlers[target.ServiceName]
	sr.mu.RUnlock()

	if !exists {
		if sr.logger != nil {
			sr.logger.Error("dispatch_unknown_service", "service_name", target.ServiceName)
		}
		return &DispatchResult{
			Success:  false,
			Error:    fmt.Sprintf("unknown service: %s", target.ServiceName),
			Duration: time.Since(startTime),
			Retries:  target.RetryCount,
		}
	}

	// Check if service is healthy
	if !svc.IsHealthy() {
		if sr.logger != nil {
			sr.logger.Error("dispatch_unhealthy_service", "service_name", target.ServiceName)
		}
		return &DispatchResult{
			Success:  false,
			Error:    fmt.Sprintf("service unhealthy: %s", target.ServiceName),
			Duration: time.Since(startTime),
			Retries:  target.RetryCount,
		}
	}

	// Check if service can accept
	if !svc.CanAccept() {
		if sr.logger != nil {
			sr.logger.Warn("dispatch_service_at_capacity", "service_name", target.ServiceName)
		}
		return &DispatchResult{
			Success:  false,
			Error:    fmt.Sprintf("service at capacity: %s", target.ServiceName),
			Duration: time.Since(startTime),
			Retries:  target.RetryCount,
		}
	}

	// Check if handler exists
	if handler == nil {
		if sr.logger != nil {
			sr.logger.Error("dispatch_no_handler", "service_name", target.ServiceName)
		}
		return &DispatchResult{
			Success:  false,
			Error:    fmt.Sprintf("no handler for service: %s", target.ServiceName),
			Duration: time.Since(startTime),
			Retries:  target.RetryCount,
		}
	}

	// Update load
	if !sr.IncrementLoad(target.ServiceName) {
		return &DispatchResult{
			Success:  false,
			Error:    fmt.Sprintf("failed to acquire load slot: %s", target.ServiceName),
			Duration: time.Since(startTime),
			Retries:  target.RetryCount,
		}
	}
	defer sr.DecrementLoad(target.ServiceName)

	// Apply timeout
	if target.TimeoutSeconds > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, time.Duration(target.TimeoutSeconds)*time.Second)
		defer cancel()
	}

	if sr.logger != nil {
		sr.logger.Debug("dispatch_started",
			"service_name", target.ServiceName,
			"method", target.Method,
		)
	}

	// Execute handler
	result, err := handler(ctx, target, data)
	if err != nil {
		// Check if context timed out
		if ctx.Err() == context.DeadlineExceeded {
			if sr.logger != nil {
				sr.logger.Error("dispatch_timeout",
					"service_name", target.ServiceName,
					"timeout_seconds", target.TimeoutSeconds,
				)
			}
			return &DispatchResult{
				Success:  false,
				Error:    "dispatch timeout",
				Duration: time.Since(startTime),
				Retries:  target.RetryCount,
			}
		}

		// Check if we should retry
		if target.CanRetry() {
			target.IncrementRetry()
			if sr.logger != nil {
				sr.logger.Info("dispatch_retry",
					"service_name", target.ServiceName,
					"retry", target.RetryCount,
					"max_retries", target.MaxRetries,
				)
			}
			return sr.Dispatch(ctx, target, data)
		}

		if sr.logger != nil {
			sr.logger.Error("dispatch_error",
				"service_name", target.ServiceName,
				"error", err.Error(),
			)
		}
		return &DispatchResult{
			Success:  false,
			Error:    err.Error(),
			Duration: time.Since(startTime),
			Retries:  target.RetryCount,
		}
	}

	if result == nil {
		result = &DispatchResult{Success: true}
	}
	result.Duration = time.Since(startTime)
	result.Retries = target.RetryCount

	return result
}

// =============================================================================
// Statistics
// =============================================================================

// GetStats returns statistics about the service registry.
func (sr *ServiceRegistry) GetStats() map[string]any {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	totalLoad := 0
	totalCapacity := 0
	healthyCount := 0
	unhealthyCount := 0

	for _, svc := range sr.services {
		totalLoad += svc.CurrentLoad
		totalCapacity += svc.MaxConcurrent
		if svc.IsHealthy() {
			healthyCount++
		} else {
			unhealthyCount++
		}
	}

	return map[string]any{
		"total_services":    len(sr.services),
		"healthy_services":  healthyCount,
		"unhealthy_services": unhealthyCount,
		"total_handlers":    len(sr.handlers),
		"total_load":        totalLoad,
		"total_capacity":    totalCapacity,
	}
}

// GetServiceStats returns per-service statistics.
func (sr *ServiceRegistry) GetServiceStats() map[string]map[string]any {
	sr.mu.RLock()
	defer sr.mu.RUnlock()

	stats := make(map[string]map[string]any, len(sr.services))
	for name, svc := range sr.services {
		_, hasHandler := sr.handlers[name]
		stats[name] = map[string]any{
			"type":            svc.ServiceType,
			"status":          string(svc.Status),
			"current_load":    svc.CurrentLoad,
			"max_concurrent":  svc.MaxConcurrent,
			"has_handler":     hasHandler,
			"last_health_check": svc.LastHealthCheck.Format(time.RFC3339),
		}
	}
	return stats
}

File: src/kernel/services.rs
//! Service registry - IPC coordinator and service discovery.
//!
//! Features:
//!   - Service registration and discovery
//!   - Health tracking
//!   - Load balancing

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// =============================================================================
// Service Status
// =============================================================================

/// Health status of a service.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ServiceStatus {
    /// Service is healthy and accepting requests
    Healthy,
    /// Service is running but with reduced capacity
    Degraded,
    /// Service is not accepting requests
    Unhealthy,
    /// Service status is unknown
    Unknown,
}

// =============================================================================
// Service Types
// =============================================================================

/// Service type constants for classifying services.
pub const SERVICE_TYPE_FLOW: &str = "flow";
pub const SERVICE_TYPE_WORKER: &str = "worker";
pub const SERVICE_TYPE_INFERENCE: &str = "inference";

// =============================================================================
// Service Info
// =============================================================================

/// ServiceInfo describes a registered service.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceInfo {
    // Identity
    pub name: String,
    pub service_type: String, // "flow", "worker", "inference"
    pub version: String,

    // Capabilities
    pub capabilities: Vec<String>,

    // Capacity
    pub max_concurrent: i32,
    pub current_load: i32,

    // Health
    pub status: ServiceStatus,
    pub last_health_check: DateTime<Utc>,

    // Metadata
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl ServiceInfo {
    /// Create a new service descriptor.
    pub fn new(name: String, service_type: String) -> Self {
        Self {
            name,
            service_type,
            version: "1.0.0".to_string(),
            capabilities: Vec::new(),
            max_concurrent: 10,
            current_load: 0,
            status: ServiceStatus::Healthy,
            last_health_check: Utc::now(),
            metadata: HashMap::new(),
        }
    }

    /// Check if the service can accept more load.
    pub fn can_accept(&self) -> bool {
        self.status == ServiceStatus::Healthy && self.current_load < self.max_concurrent
    }

    /// Check if the service is healthy.
    pub fn is_healthy(&self) -> bool {
        matches!(
            self.status,
            ServiceStatus::Healthy | ServiceStatus::Degraded
        )
    }
}

// =============================================================================
// Service Registry
// =============================================================================

/// ServiceRegistry manages service registration, discovery, and dispatch.
///
/// Single-actor implementation for tracking registered services.
#[derive(Debug)]
pub struct ServiceRegistry {
    /// Service registry
    services: HashMap<String, ServiceInfo>,
}

impl ServiceRegistry {
    /// Create a new service registry.
    pub fn new() -> Self {
        Self {
            services: HashMap::new(),
        }
    }

    // =============================================================================
    // Service Registration
    // =============================================================================

    /// Register a service with the registry.
    /// Returns true if registration succeeded, false if service already exists.
    pub fn register_service(&mut self, info: ServiceInfo) -> bool {
        if self.services.contains_key(&info.name) {
            return false;
        }

        self.services.insert(info.name.clone(), info);
        true
    }

    /// Unregister a service.
    /// Returns true if unregistration succeeded.
    pub fn unregister_service(&mut self, service_name: &str) -> bool {
        if !self.services.contains_key(service_name) {
            return false;
        }

        self.services.remove(service_name);
        true
    }

    // =============================================================================
    // Service Discovery
    // =============================================================================

    /// Get a service by name.
    pub fn get_service(&self, service_name: &str) -> Option<ServiceInfo> {
        self.services.get(service_name).cloned()
    }

    /// List all registered services matching criteria.
    pub fn list_services(
        &self,
        service_type: Option<&str>,
        healthy_only: bool,
    ) -> Vec<ServiceInfo> {
        self.services
            .values()
            .filter(|svc| {
                if let Some(stype) = service_type {
                    if svc.service_type != stype {
                        return false;
                    }
                }
                if healthy_only && !svc.is_healthy() {
                    return false;
                }
                true
            })
            .cloned()
            .collect()
    }

    /// Get all service names.
    pub fn get_service_names(&self) -> Vec<String> {
        self.services.keys().cloned().collect()
    }

    /// Check if a service is registered.
    pub fn has_service(&self, service_name: &str) -> bool {
        self.services.contains_key(service_name)
    }

    // =============================================================================
    // Load Management
    // =============================================================================

    /// Increment load for a service.
    pub fn increment_load(&mut self, service_name: &str) -> bool {
        if let Some(service) = self.services.get_mut(service_name) {
            service.current_load += 1;
            true
        } else {
            false
        }
    }

    /// Decrement load for a service.
    pub fn decrement_load(&mut self, service_name: &str) -> bool {
        if let Some(service) = self.services.get_mut(service_name) {
            service.current_load = (service.current_load - 1).max(0);
            true
        } else {
            false
        }
    }

    /// Get current load for a service.
    pub fn get_load(&self, service_name: &str) -> i32 {
        self.services
            .get(service_name)
            .map(|s| s.current_load)
            .unwrap_or(0)
    }

    // =============================================================================
    // Health Management
    // =============================================================================

    /// Update health status for a service.
    pub fn update_health(&mut self, service_name: &str, status: ServiceStatus) -> bool {
        if let Some(service) = self.services.get_mut(service_name) {
            service.status = status;
            service.last_health_check = Utc::now();
            true
        } else {
            false
        }
    }

    /// Get count of healthy services.
    pub fn get_healthy_count(&self) -> usize {
        self.services.values().filter(|s| s.is_healthy()).count()
    }

    // =============================================================================
    // Statistics
    // =============================================================================

    /// Get statistics for a specific service.
    pub fn get_service_stats(&self, service_name: &str) -> Option<ServiceStats> {
        self.services.get(service_name).map(|svc| ServiceStats {
            name: svc.name.clone(),
            service_type: svc.service_type.clone(),
            status: svc.status,
            current_load: svc.current_load,
            max_concurrent: svc.max_concurrent,
            utilization: if svc.max_concurrent > 0 {
                (svc.current_load as f64 / svc.max_concurrent as f64) * 100.0
            } else {
                0.0
            },
        })
    }

    /// Get overall registry statistics.
    pub fn get_stats(&self) -> RegistryStats {
        let mut stats = RegistryStats {
            total_services: self.services.len(),
            healthy_services: 0,
            degraded_services: 0,
            unhealthy_services: 0,
            total_load: 0,
            total_capacity: 0,
            services_by_type: HashMap::new(),
        };

        for svc in self.services.values() {
            match svc.status {
                ServiceStatus::Healthy => stats.healthy_services += 1,
                ServiceStatus::Degraded => stats.degraded_services += 1,
                ServiceStatus::Unhealthy => stats.unhealthy_services += 1,
                ServiceStatus::Unknown => {}
            }

            stats.total_load += svc.current_load as usize;
            stats.total_capacity += svc.max_concurrent as usize;

            *stats
                .services_by_type
                .entry(svc.service_type.clone())
                .or_insert(0) += 1;
        }

        stats
    }

}

impl Default for ServiceRegistry {
    fn default() -> Self {
        Self::new()
    }
}

// =============================================================================
// Statistics Types
// =============================================================================

/// Statistics for a specific service.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceStats {
    pub name: String,
    pub service_type: String,
    pub status: ServiceStatus,
    pub current_load: i32,
    pub max_concurrent: i32,
    pub utilization: f64, // Percentage
}

/// Overall registry statistics.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegistryStats {
    pub total_services: usize,
    pub healthy_services: usize,
    pub degraded_services: usize,
    pub unhealthy_services: usize,
    pub total_load: usize,
    pub total_capacity: usize,
    pub services_by_type: HashMap<String, usize>,
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_register_service() {
        let mut registry = ServiceRegistry::new();

        let service = ServiceInfo::new("flow_service".to_string(), SERVICE_TYPE_FLOW.to_string());
        let success = registry.register_service(service.clone());
        assert!(success);

        // Verify service is registered
        let retrieved = registry.get_service("flow_service");
        assert!(retrieved.is_some());
        assert_eq!(retrieved.unwrap().name, "flow_service");

        // Try to register again - should fail
        let duplicate = registry.register_service(service);
        assert!(!duplicate);
    }

    #[test]
    fn test_unregister_service() {
        let mut registry = ServiceRegistry::new();

        let service = ServiceInfo::new("flow_service".to_string(), SERVICE_TYPE_FLOW.to_string());
        registry.register_service(service);

        // Verify service exists
        assert!(registry.has_service("flow_service"));

        // Unregister
        let success = registry.unregister_service("flow_service");
        assert!(success);

        // Verify service is gone
        assert!(!registry.has_service("flow_service"));

        // Try to unregister again - should fail
        let not_found = registry.unregister_service("flow_service");
        assert!(!not_found);
    }

    #[test]
    fn test_get_service() {
        let mut registry = ServiceRegistry::new();

        let service = ServiceInfo::new("worker1".to_string(), SERVICE_TYPE_WORKER.to_string());
        registry.register_service(service);

        // Get existing service
        let retrieved = registry.get_service("worker1");
        assert!(retrieved.is_some());
        assert_eq!(retrieved.unwrap().service_type, SERVICE_TYPE_WORKER);

        // Get non-existent service
        let not_found = registry.get_service("nonexistent");
        assert!(not_found.is_none());
    }

    #[test]
    fn test_list_services_filtered() {
        let mut registry = ServiceRegistry::new();

        // Register multiple services of different types
        let flow1 = ServiceInfo::new("flow1".to_string(), SERVICE_TYPE_FLOW.to_string());
        let flow2 = ServiceInfo::new("flow2".to_string(), SERVICE_TYPE_FLOW.to_string());
        let worker1 = ServiceInfo::new("worker1".to_string(), SERVICE_TYPE_WORKER.to_string());

        registry.register_service(flow1);
        registry.register_service(flow2);
        registry.register_service(worker1);

        // Mark one as unhealthy
        registry.update_health("flow2", ServiceStatus::Unhealthy);

        // List all services
        let all = registry.list_services(None, false);
        assert_eq!(all.len(), 3);

        // List only flow services
        let flows = registry.list_services(Some(SERVICE_TYPE_FLOW), false);
        assert_eq!(flows.len(), 2);

        // List only healthy services
        let healthy = registry.list_services(None, true);
        assert_eq!(healthy.len(), 2);

        // List healthy flow services
        let healthy_flows = registry.list_services(Some(SERVICE_TYPE_FLOW), true);
        assert_eq!(healthy_flows.len(), 1);
        assert_eq!(healthy_flows[0].name, "flow1");
    }

    #[test]
    fn test_increment_decrement_load() {
        let mut registry = ServiceRegistry::new();

        let service = ServiceInfo::new("worker1".to_string(), SERVICE_TYPE_WORKER.to_string());
        registry.register_service(service);

        // Initial load is 0
        assert_eq!(registry.get_load("worker1"), 0);

        // Increment load
        let success = registry.increment_load("worker1");
        assert!(success);
        assert_eq!(registry.get_load("worker1"), 1);

        // Increment again
        registry.increment_load("worker1");
        assert_eq!(registry.get_load("worker1"), 2);

        // Decrement load
        let success = registry.decrement_load("worker1");
        assert!(success);
        assert_eq!(registry.get_load("worker1"), 1);

        // Decrement to zero
        registry.decrement_load("worker1");
        assert_eq!(registry.get_load("worker1"), 0);

        // Decrement below zero (should stay at 0)
        registry.decrement_load("worker1");
        assert_eq!(registry.get_load("worker1"), 0);

        // Try to increment non-existent service
        let not_found = registry.increment_load("nonexistent");
        assert!(!not_found);
    }

    #[test]
    fn test_update_health() {
        let mut registry = ServiceRegistry::new();

        let service = ServiceInfo::new("worker1".to_string(), SERVICE_TYPE_WORKER.to_string());
        registry.register_service(service);

        // Initial status is healthy
        let svc = registry.get_service("worker1").unwrap();
        assert_eq!(svc.status, ServiceStatus::Healthy);

        // Update to degraded
        let success = registry.update_health("worker1", ServiceStatus::Degraded);
        assert!(success);
        assert_eq!(
            registry.get_service("worker1").unwrap().status,
            ServiceStatus::Degraded
        );

        // Update to unhealthy
        registry.update_health("worker1", ServiceStatus::Unhealthy);
        assert_eq!(
            registry.get_service("worker1").unwrap().status,
            ServiceStatus::Unhealthy
        );

        // Try to update non-existent service
        let not_found = registry.update_health("nonexistent", ServiceStatus::Healthy);
        assert!(!not_found);
    }

    #[test]
    fn test_get_stats() {
        let mut registry = ServiceRegistry::new();

        // Register multiple services
        let flow1 = ServiceInfo::new("flow1".to_string(), SERVICE_TYPE_FLOW.to_string());
        let mut flow2 = ServiceInfo::new("flow2".to_string(), SERVICE_TYPE_FLOW.to_string());
        flow2.current_load = 3;
        flow2.max_concurrent = 5;

        let mut worker1 = ServiceInfo::new("worker1".to_string(), SERVICE_TYPE_WORKER.to_string());
        worker1.current_load = 2;
        worker1.max_concurrent = 10;

        registry.register_service(flow1);
        registry.register_service(flow2);
        registry.register_service(worker1);

        // Update health statuses
        registry.update_health("flow2", ServiceStatus::Degraded);

        // Get overall stats
        let stats = registry.get_stats();
        assert_eq!(stats.total_services, 3);
        assert_eq!(stats.healthy_services, 2);
        assert_eq!(stats.degraded_services, 1);
        assert_eq!(stats.unhealthy_services, 0);
        assert_eq!(stats.total_load, 5); // 0 + 3 + 2
        assert_eq!(stats.total_capacity, 25); // 10 + 5 + 10
        assert_eq!(stats.services_by_type.get(SERVICE_TYPE_FLOW), Some(&2));
        assert_eq!(stats.services_by_type.get(SERVICE_TYPE_WORKER), Some(&1));

        // Get service-specific stats
        let flow2_stats = registry.get_service_stats("flow2").unwrap();
        assert_eq!(flow2_stats.name, "flow2");
        assert_eq!(flow2_stats.current_load, 3);
        assert_eq!(flow2_stats.max_concurrent, 5);
        assert_eq!(flow2_stats.utilization, 60.0); // 3/5 * 100
    }

    #[test]
    fn test_service_can_accept() {
        let mut service = ServiceInfo::new("test".to_string(), SERVICE_TYPE_FLOW.to_string());
        service.max_concurrent = 5;
        service.current_load = 3;

        // Healthy and below capacity
        assert!(service.can_accept());

        // At capacity
        service.current_load = 5;
        assert!(!service.can_accept());

        // Unhealthy
        service.current_load = 3;
        service.status = ServiceStatus::Unhealthy;
        assert!(!service.can_accept());

        // Degraded but below capacity (should accept)
        service.status = ServiceStatus::Degraded;
        assert!(service.is_healthy()); // Degraded is considered healthy for is_healthy()
    }
}
